### DISCUSSION TITLE
Formal verification of Clarity based smart contracts

### DISCUSSION THREAD

**Discussion by hassan-truscova** [stacks-network/stacks-core#5615] [Clarity]
Hi Everyone,
My name is [Hassan ](https://truscova.com/index.php#TEAM) from [Truscova GmbH](https://truscova.com/).
Recently, I was looking at the verification/testing tooling for Clarity based smart contracts, in particular related to fuzzing, symbolic execution, and formal verification (model checking, theorem proving) etc, however, i could not find much.

The good thing is that i got in contact with @moodmosaic via his [Blog article](https://blog.nikosbaxevanis.com/2023/01/30/invariant-testing-from-command-line/). He was kind enough to respond to my emails and for a very helpful discussion. He also introduced me to [rendezvous](https://github.com/stacks-network/rendezvous) for fuzzing the heck out of the Clarity-based smart contracts, which is pretty cool.

I wanted to know if there is any interest or if there are any tools for doing formal verification of Clarity-based smart contracts? or if anyone is already doing it?
Any leads, contacts, connections would be really helpful :)

Thank you very much in advance.

**Comment by jcnelson** [comment]
Hi there!  I see you've been in touch with @moodmosaic elsewhere.

I think there is interest in formal verification, and as far as I know, no one is doing it yet.  This is currently impeded by the lack of a formal language specification.  AFAIK, there isn't even a grammar written down yet.  But because Clarity is a (comparatively speaking) very simple language, it shouldn't be an insurmountable task to write a correct one; it's just that the core blockchain devs have been tied up until recently with shipping and stabilizing Nakamoto.

If this is something you'd like to pursue, the best people to talk to are @kantai and @obycode.

**Reply by hassan-truscova** [reply]
Hi @jcnelson, thank you very much for the response. If there is interest in formal verification, i would be very interested in knowing more about what opportunities exist. Do i need to send PMs to @kantai and @obycode ?

**Comment by moodmosaic** [comment]
This could be valuable for fuzzing tools like rendezvous. While rendezvous tests properties probabilistically, formal verification would check all paths exhaustively. Another use case would be enabling systematic reasoning about AST nodes, making it easier to compute min/max runtime costs, as discussed in https://github.com/stacks-network/stacks-core/issues/5360#issuecomment-2430355159. There are likely other use cases worth exploring too.

**Comment by atifhass** [comment]
Hello @hassan-truscova @moodmosaic @kantai @obycode 

We've developed some tools that we’re currently testing out internally and with some friendly users that are specifically addressing formal verification and audits in general. We’ve developed two tools:

- **Formal Verification Tool  - A SMT based Bounded Model Checking(BMC) engine 
- AI Audit Agent - Smart Code Scan powered by leveraging fine-tuned LLMs which scans for vulnerabilities and potential bugs**

Here’s a link to our demo youtube video. [link](https://www.youtube.com/watch?v=GbzLhrrZkv8&t=47s) 

We'd love to hear your feedback and offer a live demo to test these features. Let us know how you would like us to set up your accounts to try out the tools!

Learn more about Clara at - https://www.goclara.ai/

**Reply by hassan-truscova** [reply]
@atifhass the youtube demo looks cool. I am unsure about what is happening on the formal engine side. Would it be possible to schedule a demo? do i need to schedule it on the website?

**Reply by moodmosaic** [reply]
I’d be interested too in that demo, if that’s possible.

**Reply by radubahmata** [reply]
If a live demo happens, I'd be interested in it.

**Reply by atifhass** [reply]
@hassan-truscova @moodmosaic @BowTiedRadone for sure we can setup a live demo. Let me know what time slot works best for you guys, we can work out a slot for next week Wednesday(12th march) or Thursday or Friday, between 9 pm - 12 am (GMT+5). Looking forward.

**Reply by hassan-truscova** [reply]
@atifhass @moodmosaic @BowTiedRadone Does Wednesday, 12.03.2025 at 9 PM (GMT+5) work for all of you? that would be 10 PM CEST, if i am not not wrong.

**Reply by moodmosaic** [reply]
Thanks, works for me.  

@atifhass, can you run the tool on this contract for the demo?  https://gist.github.com/moodmosaic/2079d9d9a0d230dbcbcba2b8ff1e4cf8

Would be great to see actual results and measure performance.

**Reply by atifhass** [reply]
Can you guys please drop your email addresses here, will send out a calendar invite (once @BowTiedRadone also confirms). 

@moodmosaic let me share back the audit report with you.

**Reply by hassan-truscova** [reply]
@atifhass i won't drop my email id here, bots will spam my email ;). please visit our [website](https://truscova.com/) and scroll to the bottom, you will find email id.

**Reply by atifhass** [reply]
@hassan-truscova dropping calendly link with the date/time slot we finalised. 

https://calendly.com/atif-hassan-ttjm/clara

**Reply by hassan-truscova** [reply]
@atifhass thanks. I have registered at the calendly link. It shows 17:00 Hrs CEST (my time zone) but it works for me.

**Reply by moodmosaic** [reply]
@atifhass, it shows no available hours. Can’t get in. My email is in my profile.

**Reply by atifhass** [reply]
@moodmosaic adding you to the event

**Reply by radubahmata** [reply]
@atifhass works for me!

**Reply by radubahmata** [reply]
Got the invite. Thank you! Looking forward for the demo! 🚀

**Reply by hassan-truscova** [reply]
@atifhass @BowTiedRadone @moodmosaic thank you very much for attending and presenting the demo. For me personally, it is always nice to see formal verification being used in industry :) Having said that, @atifhass, during the demo your colleague mentioned that maybe you will make the reports publicly available. Is there any timeline for that? 

Secondly, given that there is a fuzzing tool (Rendezvous) already available, is it possible for you guys to make a sliced or limited version of your formal engine available so that i can also do some benchmarking?

third, you mentioned during the demo that you are running into some issues related to mappings and iterative structures. Would you be open to have another discussion? maybe i can help you guys there.

Looking forward.

**Reply by mali-tintash** [reply]
@hassan-truscova ,
@moodmosaic @BowTiedRadone and Rapha. Thank you guys for taking the time out for the demo presentation.

@hassan-truscova , regarding issues related to mappings of iterative structures @atifhass will coordinate with you . Thank you for agreeing to help us around that.
We are in the process of making it CLI friendly. We'd be happy to share it with you for feedback and improvements.

The reports and the target file are attached below
[Audit Report.pdf](https://github.com/user-attachments/files/19225750/Audit.Report.pdf)
[Formal Verification Report.pdf](https://github.com/user-attachments/files/19225753/Formal.Verification.Report.pdf)
[target_original.zip](https://github.com/user-attachments/files/19225789/target_original.zip)
